### ISSUE TITLE
Error mangling in an error message for typed holes

### ISSUE THREAD

**Issue by codeworld-feedback** [google/codeworld#1097]
**Program:** https://code.world/#PQi2vxqOWd0Tle1fcO6Y-9Q

**Comment:**

This error message has something mangled in it.

**Message:**

    program.hs:47:23: warning:
        • Found hole: _ :: (Number, Number, Number, Number) -> Truth
        • In the expression: _ (ox, oy, x, y)
          In the argument(s) of any, namely
            ([_ (ox, oy, x, y) |
                 (ox, _) <- first (obs, 10) |  oy <- [0, 4 .. ]])
        • Valid hole fits include
            collidesObstacle :: (Number, Number, Number, Number) -> Truth
              (bound at program.hs:51:1)
            undefined :: forall a.
                         
    Truth)
              (defined in the standard library
               (and originally defined in the standard library))

**Comment by cdsmith** [comment]
Current simplest reproduction: https://code.world/#PQ9vOLk3xExhvQsFKyza3jA

Here's the original GHC error message:

```
Test.hs:1:8: error:
    • Found hole: _ :: t -> t1
      Where: ‘t’, ‘t1’ are rigid type variables bound by
               the inferred type of f :: t -> t1
               at Test.hs:1:1-11
    • In the expression: _
      In the expression: _ (x)
      In an equation for ‘f’: f (x) = _ (x)
    • Relevant bindings include
        x :: t (bound at Test.hs:1:3)
        f :: t -> t1 (bound at Test.hs:1:1)
      Valid hole fits include
        f :: forall t1 t2. t1 -> t2
          with f @t @t1
          (defined at Test.hs:1:1)
        undefined :: forall a.
                     base-4.12.0.0:GHC.Stack.Types.HasCallStack =>
                     a
          with undefined @(t -> t1)
          (imported from ‘Prelude’ at Test.hs:1:1
           (and originally defined in ‘codeworld-base-0.2.0.0:Internal.Prelude’))
```

Here's the mangled message:

```
program.hs:1:8: warning:
    • Found hole: _ :: t -> t1
      Where: t, t1 are rigid type variables bound by
               the inferred type of f :: t -> t1
               at program.hs:1:1-11
    • Valid hole fits include
        f :: forall t1 t2. t1 -> t2
          with f @t @t1
          (defined at program.hs:1:1)
        undefined :: forall a.
                     
t1)
          (defined in the standard library
           (and originally defined in the standard library))```
```

The GHC error message is correctly clean of "in the expression ..." context, and relevant bindings, since we don't want these.  But there are three problems with the "Valid hole fits" section:

1. We don't want the line `with f @t @t1`, since visible type application is overly technical
2. The `HasCallStack` constraint from `undefined` is removed, but over-aggressively, killing part of the following expression and the following line.
3. `(and originally defined in ...)` is unnecessary; imports are not used enough to bother.

**Comment by cdsmith** [comment]
Worst mangling is fixed in commit dd12d472903f32fa4f7c5666c44d57cc8222f608

What remains is just to strip out the unnecessary parts, of which there are actually three:

1. `Where: t, t1 are rigid type variables...`
2. `with f @t @t1` and `with undefined @(t -> t1)`
3. `(and originally defined in the standard library)`
